PluginsAndFeatures/azure-toolkit-for-eclipse/com.microsoft.azuretools.container/src/com/microsoft/azuretools/container/handlers/PublishHandler.java

```diff
@@ -53,35 +65,77 @@ public Object onExecute(ExecutionEvent event) throws ExecutionException {
+            });
+        });
 
-            MessageDialog.openError(window.getShell(), "Error on building image", String.format(
-                    Constant.ERROR_BUILDING_IMAGE, dockerHost, dockerFilePath.replaceFirst("^/", ""), e.getMessage()));
-            return null;
-        }
-        PublishWizard pw = new PublishWizard();
-        WizardDialog pwd = new PublishWizardDialog(window.getShell(), pw);
-        if (pwd.open() == Window.OK) {
-            ConsoleLogger.info(String.format("URL: http://%s.azurewebsites.net/%s",
-                    DockerRuntime.getInstance().getLatestWebAppName(), project.getName()));
-            props = DockerRuntime.getInstance().saveToProps(props);
-            ConfigFileUtil.saveConfig(project, props);
-        }
         return null;
     }
 
     private void buildImage(IProject project) throws Exception {
-        DockerClient dockerClient = DockerRuntime.getInstance().getDockerBuilder().build();
-        String destinationPath = project.getLocation() + Constant.DOCKER_CONTEXT_FOLDER + project.getName() + ".war";
-        WarUtil.export(project, destinationPath);
-        DockerUtil.buildImage(dockerClient, project, project.getLocation() + Constant.DOCKER_CONTEXT_FOLDER);
-        DockerRuntime.getInstance().setLatestArtifactName(project.getName());
+        if (project == null) {
+            throw new Exception("Project is null.");
+        }
+        String basePath = project.getLocation().toString();
+        DockerClient docker = DockerRuntime.getInstance().getDockerBuilder().build();
+        // export WAR file
+
+        String destinationPath = project.getLocation() + Constant.DOCKERFILE_FOLDER + project.getName() + ".war";
+        ConsoleLogger.info(String.format(Constant.MESSAGE_EXPORTING_PROJECT, destinationPath));
+
+        if (MavenUtils.isMavenProject(project)) {
+            destinationPath = MavenUtils.getTargetPath(project);
+            MavenUtils.executePackageGoal(project);
+        } else {
+            destinationPath = Paths.get(basePath, Constant.DOCKERFILE_FOLDER, project.getName() + ".war").normalize()
+                    .toString();
+            WarUtil.export(project, destinationPath);
+        }
+
+        // validate dockerfile
+        Path targetDockerfile = Paths.get(basePath, Constant.DOCKERFILE_FOLDER, Constant.DOCKERFILE_NAME);
+        ConsoleLogger.info(String.format("Validating dockerfile ... [%s]", targetDockerfile));
+        if (!targetDockerfile.toFile().exists()) {
+            throw new FileNotFoundException("Dockerfile not found.");
+        }
+        // replace placeholder if exists
+        String content = new String(Files.readAllBytes(targetDockerfile));
+        content = content.replaceAll(Constant.DOCKERFILE_ARTIFACT_PLACEHOLDER,
+                Paths.get(basePath).toUri().relativize(Paths.get(destinationPath).toUri()).getPath());
+        Files.write(targetDockerfile, content.getBytes());
```

Comment:
This part of code is duplicate with the one in local run. Can we abstract it?

Reply:
low prior now